### ISSUE TITLE
Scripting rclone

### ISSUE THREAD

**Issue by hashbackup** [ncw/rclone#1136]
It would be really nice if rclone had more detailed exit code handling to make it easier to use for scripting or from other programs.  Just as a starting point:

0 = success
non-zero = failure

More detail about failures (bit 1 = LSB):

bit 1: 
- if 1, do not try again because it won't work, like rclone --xyz (unrecognized argument), or rclone ls filename where filename is not present on the remote
- if 0, request might work if issued again, for example, rclone was throttled every time and retries exceeded, or broken pipe (remote closed connection) and retries exceeded

bit 2-3:
- 0 = remote service was never contacted, eg, bad rclone arguments, socket timeout on connection, DNS failure mapping name to IP address
- 1 = remote service contacted, but no response from actual request, eg, throttled, 500 server error, timeout reading response from server, broken pipe.  If doing an ls command and rclone is able to get info about the drive but the ls request is never actually issued, this bit would be set, because getting drive info was not the actual request (it was an ls)
- 2 = remote service contacted, request was issued and a response received
- 3 = error relates to local, not remote

That leaves 5 bits for 32 more specific error codes:
- 0 = no specific error
- 1 = file not found
- 2 = no permission to perform operation
- 3 = quota exceeded
...

A detailed exit code lets scripts and other programs use rclone as a transport, and make decisions about what actually happened.  With just a zero/non-zero exit code, outer programs don't know if rclone failed, the remote failed, the rclone command is bad, or the request itself failed (like getting a file that doesn't exist for example).

For example, if an outer program wants to delete a file, it would be nice to distinguish between a) the file isn't there, which may or may not be treated as an error; b) the request is bad (program bug); or c) the delete request failed for some reason, but might work later, and it's unknown whether the file is there.

rsync has a detailed list of error codes if something like that would work better.  Thanks!

**Comment by ncw** [comment]
I think for rclone a list of error codes like rsync would probably work best.

Then I can assign specific errors codes..

I'd probably want to keep it simple to start with.  Rclone can classify some errorrs - these are the categories used at the moment, which I could map to exit codes...
 
  * 0 success
  * 1 Syntax or usage error
  * 2 Error not otherwise categorised
  * 3 Directory not found
  * 4 File not found
  * 5 Temporary error (one that more retries might fix) (Retry errors)
  * 6 Less serious errors (like 461 errors from dropbox) (NoRetry errors)
  * 7 Fatal error (one that more retries won't fix, like account suspended) (Fatal errors)

How would that work for you?

**Comment by hashbackup** [comment]
This would be a big improvement.  Today, it's hard to make good decisions about the state of the remote based on either exit codes or output received.  Thanks again!

**Comment by ncw** [comment]
OK the above would be relatively easy!

**Comment by kubark42** [comment]
Agreed. I'm currently scripting by regexping the output, but some kind of more direct message would be very nice.
